stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/MessageDispatcher.java

```diff
@@ -221,17 +232,69 @@ private void handleError(byte[] message, long deliveryTag, AMQP.BasicProperties
         ObjectValue messageObjectValue = getMessageObjectValue(message, deliveryTag, properties);
         CountDownLatch countDownLatch = new CountDownLatch(1);
         try {
-            runtime.invokeMethodAsync(service, RabbitMQConstants.FUNC_ON_ERROR,
-                    new RabbitMQResourceCallback(countDownLatch), messageObjectValue, true, error, true);
+            executeOnErrorResource(countDownLatch, messageObjectValue, error, properties);
             countDownLatch.await();
         } catch (InterruptedException e) {
             Thread.currentThread().interrupt();
+            RabbitMQMetricsUtil.reportError(channel, RabbitMQObservabilityConstants.ERROR_TYPE_CONSUME);
             throw new RabbitMQConnectorException(RabbitMQConstants.THREAD_INTERRUPTED);
         } catch (AlreadyClosedException | BallerinaConnectorException exception) {
+            RabbitMQMetricsUtil.reportError(channel, RabbitMQObservabilityConstants.ERROR_TYPE_CONSUME);
             throw new RabbitMQConnectorException("Error occurred in RabbitMQ service. ");
         }
     }
 
+    private void executeResource(CountDownLatch countDownLatch, byte[] message, long deliveryTag,
+                                 AMQP.BasicProperties properties) {
+        if (ObserveUtils.isTracingEnabled()) {
+            Map<String, Object> observerProperties = getNewObserverContextInProperties();
+            runtime.invokeMethodAsync(service, RabbitMQConstants.FUNC_ON_MESSAGE,
+                                      new RabbitMQResourceCallback(countDownLatch, channel, queueName, message.length),
+                                      observerProperties, getMessageObjectValue(message, deliveryTag, properties),
+                                      true);
+            return;
+        }
+        runtime.invokeMethodAsync(service, RabbitMQConstants.FUNC_ON_MESSAGE,
+                                  new RabbitMQResourceCallback(countDownLatch, channel, queueName, message.length),
+                                  getMessageObjectValue(message, deliveryTag, properties), true);
+    }
+
+    private void executeResource(CountDownLatch countDownLatch, byte[] message, ObjectValue messageObjectValue,
+                                 Object forContent) {
+        if (ObserveUtils.isTracingEnabled()) {
+            Map<String, Object> observerProperties = getNewObserverContextInProperties();
+            runtime.invokeMethodAsync(service, RabbitMQConstants.FUNC_ON_MESSAGE,
+                                      new RabbitMQResourceCallback(countDownLatch, channel, queueName, message.length),
+                                      observerProperties, messageObjectValue, true, forContent, true);
+            return;
```

Comment:
Can make these two methods more simpler. It does the same thing and runtime.invokeMethodAsync() is called 4 times inside these two methods :) I think we can make it one. Except for the countdownLatch and the RabbitMQResourceCallback, everything else goes into the Object.. args parameter of the invokeMethodAsync() method. This should give you a start to refactoring.